tensorflow/java/src/gen/cc/source_writer.cc

```diff
@@ -14,19 +14,47 @@ limitations under the License.
 ==============================================================================*/
 
 #include <string>
+#include <algorithm>
+#include <deque>
 
 #include "tensorflow/java/src/gen/cc/source_writer.h"
 
 namespace tensorflow {
+namespace java {
+namespace {
 
-SourceWriter& SourceWriter::Append(const StringPiece& str) {
-  if (!str.empty()) {
-    if (newline_) {
-      DoAppend(left_margin_ + line_prefix_);
-      newline_ = false;
-    }
-    DoAppend(str);
+template <typename TypeVisitor>
+void VisitType(const Type& type, TypeVisitor* visitor) {
```

Comment:
Can we do away with a templated definition and instead define a clear interface for type visitor? I was thinking of something like this:

```c++
class TypeVisitor {
  public:
    void Visit(const Type& type) {
       DoVisit(type);
       for (const Type& t : type.parameters()) {
          DoVisit(t);
       }
       for (const Type& t : type.annotations()) {
          DoVisit(t);
       }
       for (const Type& t : type.supertypes()) {
          DoVisit(t);
       }
     }
  protected:
    virtual void DoVisit(const Type& t) = 0;   
}
```

And then `GenericNamespace` and `ImportScanner` can implement this interface.
That makes the contract easier, avoids possible confusion with operator overloads. What do you think?

Reply:
No problem